### PR TITLE
Native Poseidon

### PR DESCRIPTION
**Pull request by mike1729**
Moves calculating poseidon hash function to native by exposing it to wasm executor.

### CODE REVIEW

**Review by deuszx — APPROVED** [review]

File: primitives/Cargo.toml [hunk 1]
```diff
@@ -16,6 +16,11 @@ sp-core = { default-features = false, git = "https://github.com/Cardinal-Cryptog
 sp-runtime = { default-features = false, git = "https://github.com/Cardinal-Cryptography/substrate.git", branch = "liminal-v0.9.32" }
 sp-std = { default-features = false, git = "https://github.com/Cardinal-Cryptography/substrate.git", branch = "liminal-v0.9.32" }
 sp-staking = { default-features = false, git = "https://github.com/Cardinal-Cryptography/substrate.git", branch = "liminal-v0.9.32" }
+sp-runtime-interface = { default-features = false, git = "https://github.com/Cardinal-Cryptography/substrate.git", branch = "liminal-v0.9.32" }
+
+ark-poseidon = { default-features = false, path="../poseidon", package = "liminal-ark-poseidon" }
```
Comment (pmikolajczyk41) [inline primitives/Cargo.toml:21]: I'm slightly against this naming, as it strongly suggests that this crate is from arkworks :| if you don't like the long version, maybe just `poseidon`?

File: pallets/baby-liminal/src/benchmarking/suite.rs [hunk 2]
```diff
@@ -20,6 +21,10 @@ fn insert_key<T: Config>(key: Vec<u8>) {
     VerificationKeys::<T>::insert(IDENTIFIER, BoundedVec::try_from(key).unwrap());
 }
 
+fn gen_input(x: u32) -> (u64, u64, u64, u64) {
```
Comment (pmikolajczyk41) [inline pallets/baby-liminal/src/benchmarking/suite.rs:24]: `gen_poseidon_host_input`?

File: primitives/src/host_functions/mod.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+fn output(field_element: ark_poseidon::Fr) -> Output {
+    let a = field_element.0 .0;
+    (a[0], a[1], a[2], a[3])
+}
+
+#[sp_runtime_interface::runtime_interface]
+pub trait Poseidon {
+    fn hash_one_to_one(input: Input) -> Output {
```
Comment (pmikolajczyk41) [inline primitives/src/host_functions/mod.rs:20]: `x_to_one_hash`? so that we stay consistent with both `poseidon` lib and wasm counterparts

**Review by pmikolajczyk41 — APPROVED** [review]
nice!

File: pallets/baby-liminal/src/weights.rs [hunk 4]
```diff
@@ -120,169 +140,198 @@ trait BenchmarkInfo {
...
+		Weight::from_ref_time(54_101_966_000_u64)
 			.saturating_add(T::DbWeight::get().reads(2_u64))
 	}
 	// Storage: BabyLiminal VerificationKeys (r:1 w:0)
 	// Storage: System ParentHash (r:1 w:0)
 	fn verify_groth16_merkle_tree_64() -> Weight {
-		// Minimum execution time: 45_142_204 nanoseconds.
-		Weight::from_ref_time(46_418_785_000_u64)
+		// Minimum execution time: 46_770_237 nanoseconds.
+		Weight::from_ref_time(55_113_514_000_u64)
```
[the remark this answers is not in the document]
Comment (mike1729) [reply pallets/baby-liminal/src/weights.rs:199]: It was run on my home laptop for web browsing. Actually, I was surprised how close the results were :P

File: primitives/Cargo.toml (hunk 1 above)
Comment (mike1729) [reply primitives/Cargo.toml:21]: I cannot call it poseidon since runtime interface creates module with the same name.

Comment (pmikolajczyk41) [reply primitives/Cargo.toml:21]: `liminal-ark-poseidon`? :P

File: primitives/src/host_functions/mod.rs (hunk 3 above)
Comment (mike1729) [reply primitives/src/host_functions/mod.rs]: done

File: primitives/Cargo.toml (hunk 1 above)
Comment (mike1729) [reply primitives/Cargo.toml:21]: I can feel that you tried to get your PR merged recently and somebody picked on you hard...
done

Comment (pmikolajczyk41) [reply primitives/Cargo.toml]: xD

**Pull request merged**